### PR TITLE
Route blinding: support forwarding as the intro node

### PR DESCRIPTION
**Pull request by valentinewallace**
Adds support for being the intro node to a blinded path. As such, we won't advertise route blinding support in our feature bits yet. 

The next PR (#2688) adds support for receiving to a multi-hop blinded path, to be followed by complete blinded forwarding support. #2688 also adds testing for the success case of forwarding-as-intro.

Helps address #1970. 

~Based on #2514, #2413~

### CODE REVIEW

File: lightning/src/ln/channelmanager.rs
Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs:136]: Do you mean everywhere, i.e. in `HTLCPreviousHopData`? I'd like to keep that as an `Option` for backwards compat, but I'll convert the inner bool into an enum if that sounds good.

**Comment by valentinewallace** [comment]
Expanded test coverage a bit and broke out a few more commits. Think this is ready for review.

File: lightning/src/ln/channelmanager.rs
Comment (jkczyz) [inline lightning/src/ln/channelmanager.rs:2973]: nit: no need for linking brackets in non-rustdoc comments

File: lightning/src/ln/channelmanager.rs
Comment (jkczyz) [inline lightning/src/ln/channelmanager.rs:3200]: Maybe name this something like `amount_and_cltv`? Or maybe use `map_err()?` instead of `match`?

**Review by jkczyz — COMMENTED** [review]
Still working through some of the commits

File: lightning/src/ln/channelmanager.rs
Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs:117]: Or does that make consuming this difficult?

File: lightning/src/ln/channelmanager.rs
Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs:208]: `BlindType` (or `Category`)? `Origin` sounds like the sender to me.

File: lightning/src/ln/channel.rs
Comment (valentinewallace) [reply lightning/src/ln/channel.rs:6989]: Good point, it's not required anymore since I switched these to odd TLVs. Still, it is nice to not write a bunch of `None`s in 99% of cases? 

Matt had some feedback in the past that we should macro-ize ser code like this, since it's also repeated in `Route::write`. Could make a follow-up for that, as another option.

File: lightning/src/ln/channelmanager.rs
Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs:117]: It'd be more verbose to consume and we'd have to break out of the serialization macro if we wanted to maintain backwards compat. Would also then want to add `BlindedReceive and `BlindedReceiveKeysend`, potentially. I kinda think it's fine as an optional field but open to other arguments.

**Comment by valentinewallace** [comment]
Addressed feedback.

File: lightning/src/ln/channelmanager.rs
Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs]: Maybe `BlindedErrorPolicy::{Non}IntroductionNode`? Like it refers to the plan for how we'll error if the htlc fails? Bit long but...

Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs]: Will it basically be `BlindedForward` without any data? I was about to suggest that name but it is taken lol. Would a `bool` suffice? Or at least a name similar to `BlindedForward` but for failing backwards?

Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs]: I had it as a bool previously but Matt wanted an enum because we'd have to store it in `HTLCPreviousHopData` as `blinded: Option<bool>` which is a bit weird. 

`BlindedForwardFailure`? Except we'll use it for received htlcs too. `BlindedFailureType`?

Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs]: Discussed offline, going with `BlindedFailure::From{Non}IntroductionNode`.

File: lightning/src/ln/shutdown_tests.rs
Comment (jkczyz) [inline lightning/src/ln/shutdown_tests.rs:408]: Why test this change in `shutdown_tests`?

File: lightning/src/ln/blinded_payment_tests.rs
Comment (jkczyz) [inline lightning/src/ln/blinded_payment_tests.rs:392]: Don't need to clone since `chan_upd` is never used later.

File: lightning/src/ln/blinded_payment_tests.rs
Comment (jkczyz) [inline lightning/src/ln/blinded_payment_tests.rs:420]: Indentation is off

**Review by jkczyz — COMMENTED** [review]
Alright, thanks for the patience on this. Finally was able to sit down and review all the tests. No major concerns. Feel free to squash.

File: lightning/src/ln/shutdown_tests.rs
Comment (valentinewallace) [reply lightning/src/ln/shutdown_tests.rs:408]: Just to keep the tests DRY, it's a somewhat long test.

**Review by jkczyz — COMMENTED** [review]
LGTM, though @TheBlueMatt should take a look, too.

**Comment by valentinewallace** [comment]
Squashed.

File: lightning/src/ln/channel.rs
Comment (valentinewallace) [reply lightning/src/ln/channel.rs:7016]: I'm pretty sure you're right, which unfortunately means we have an existing bug in the skimmed fees serialization :( will try to have a fix and test up soon.

File: lightning/src/ln/onion_utils.rs
Comment (valentinewallace) [reply lightning/src/ln/onion_utils.rs:440]: Done. Could also make the vec pass-by-value instead.

File: lightning/src/ln/channel.rs
Comment (valentinewallace) [reply lightning/src/ln/channel.rs]: Fixed and added a ser test.

File: lightning/src/ln/msgs.rs
Comment (valentinewallace) [reply lightning/src/ln/msgs.rs:2364]: Discussed offline and did a bit of scoping for fixing the timing oracle. Seems like the plan is to more urgently prioritize this refactor but it won't hold up the outstanding route blinding PRs.

**Comment by valentinewallace** [comment]
Rebased due to conflict.

**Review by jkczyz — COMMENTED** [review]
Looks like this needs another rebase

**Comment by valentinewallace** [comment]
Rebased pt 2.

File: lightning/src/ln/channelmanager.rs
Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs]: Yeah good point, I just squashed this fix in.

File: lightning/src/ln/channelmanager.rs
Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs:4184]: We will always fail to decode `onion_packet` if it's a phantom, because we use `Recipient::Node` instead of `::PhantomNode` when decoding the encrypted TLVs. Not much we can do here until we add full phantom support, IIUC.

File: lightning/src/ln/channelmanager.rs
Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs]: Doesn't reduce the diff but this is def cleaner and I actually meant to do it but forgot 🤦‍♀️ thanks

File: lightning/src/ln/channelmanager.rs
Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs]: I'm unsure whether it's best to blind errors before returning them here. If people are using this API as part of HTLC interception, then CLN/LND should know to blind the error on their behalf anyway, and the unblinded error could be useful for debugging. Not sure about the LSPS4 use case.

Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs]: Ok I added this to the BOLT 12 tracking issue. I don't feel super strongly so we can go ahead and plan to blind the error here.

**Comment by valentinewallace** [comment]
Noted the follow-ups here: https://github.com/lightningdevkit/rust-lightning/pull/2540#pullrequestreview-1754409841 in #1970.

**Review by jkczyz — APPROVED** [review]

**Pull request merged**
